### PR TITLE
Add metric to search and compare endpoints

### PR DESCRIPTION
**Pull request by kevinschaper**

### CODE REVIEW

**Comment by glass-ships** [comment]
Fails to install due to library linking?
```
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
```

Sorry to ping you on this @hrshdhgd but do you have any suggestions for how to resolve?
I ran a command suggested in a stack overflow post and got the following output:
```
$ cargo clean && cargo build -vv 2>/dev/null | grep 'rustc-link-lib'
[generator 0.7.5] cargo:rustc-link-lib=static=asm
[libsqlite3-sys 0.27.0] cargo:rustc-link-lib=static=sqlite3
```
but am unsure where to go from here

**Comment by justaddcoffee** [comment]
@glass-ships @kevinschaper what about [this](https://github.com/monarch-initiative/semsimian-server/pull/13)

this seems to fix the GH failure

File: src/bin.rs [hunk 1]
```diff
@@ -7,8 +7,8 @@ use semsimian_server::{compare_termsets, search, say_hello};
 
 #[launch]
 pub fn rocket() -> _ {
-    compare_termsets("HP:0000001,HP:0000002", "HP:0000003,HP:0000004");
-    search("HP:0000001,HP:0000002", "ZFIN", Some(1));
+    compare_termsets("HP:0000001,HP:0000002", "HP:0000003,HP:0000004", "AncestorInformationContent");
+    search("HP:0000001,HP:0000002", "ZFIN", "AncestorInformationContent", Some(1));
```
Comment (hrshdhgd) [inline src/bin.rs:11]: This needs to be `MetricEnum::AncestorInformationContent`

**Comment by hrshdhgd** [comment]
Thanks @justaddcoffee !

**Comment by glass-ships** [comment]
interesting, thanks for taking a look! 
out of curiosity, how would I make the same fix locally? I'm running into the same build error on my machine as well

**Comment by glass-ships** [comment]
oh weiiird, still failing here... maybe something about whatever we changed in this PR has changed the requirements behind the scenes?

**Comment by hrshdhgd** [comment]
let me make a rc1 release and lets see if that works ...my guess is installing from GitHub may be the cause

**Comment by hrshdhgd** [comment]
[v0.2.14-rc1](https://crates.io/crates/semsimian/0.2.14-rc1)

**Comment by glass-ships** [comment]
nice, thanks for getting that out so quickly! 
unfortunately same thing is happening

**Comment by hrshdhgd** [comment]
ok @glass-ships , @justaddcoffee , @kevinschaper @caufieldjh when I run `cargo test` locally , both tests pass. Can any (or all) of you confirm if it passes for you all locally too?

**Comment by justaddcoffee** [comment]
yes, `cargo test` passes for me on 5c0f8655260de20ecfe664a09dcbe8744fe34d2d (on my M1 Mac)

**Comment by caufieldjh** [comment]
I'm still getting the linking error on my system (Ubuntu 22.04.4 on WSL2) as of 7a34b35d99ba294e0c355f727318273e00304235

i.e. this part is the same as on GH actions:
```
  = note: /usr/bin/ld: /home/harry/semsimian-server/target/debug/deps/libpyo3-d15e9e6450028332.rlib(pyo3-d15e9e6450028332.pyo3.e5f8a9c18bc44ae7-cgu.8.rcgu.o): in function `pyo3_ffi::object::Py_DECREF':
          /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-ffi-0.20.3/src/object.rs:606: undefined reference to `_Py_Dealloc'
          collect2: error: ld returned 1 exit status

  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `semsimian_server` (bin "semserver") due to previous error
```

**Comment by hrshdhgd** [comment]
ok, @caufieldjh what does your local system `rustup show` yield?
here's mine: 
```
Default host: x86_64-apple-darwin
rustup home: ~/.rustup

stable-x86_64-apple-darwin (default)
rustc 1.76.0 (07dca489a 2024-02-04)
```

**Comment by hrshdhgd** [comment]
Also could you set your env variable `PYO3_PRINT_CONFIG=1` and run again. The error may be more detailed

**Comment by caufieldjh** [comment]
oh I am using _ancient_ rustc:
```
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/harry/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu (default)
nightly-x86_64-unknown-linux-gnu

active toolchain
----------------

stable-x86_64-unknown-linux-gnu (default)
rustc 1.71.1 (eb26296b5 2023-08-03)
```

and my pyO3 print config:
```
  implementation=CPython
  version=3.10
  shared=true
  abi3=false
  lib_name=python3.10
  lib_dir=/usr/lib/x86_64-linux-gnu
  executable=/usr/bin/python
  pointer_width=64
  build_flags=
  suppress_build_script_link_lines=false
```

**Comment by caufieldjh** [comment]
Still getting the error with 
```
stable-x86_64-unknown-linux-gnu (default)
rustc 1.76.0 (07dca489a 2024-02-04)
```

**Comment by justaddcoffee** [comment]
One brief commment here - notice above that 3638cf2 passes and 2-3 commits later 5e9d5e1 fails. 

So adding back one-by-one whatever changes you get when you do `git diff 5e9d5e1 3638cf2` might be a way to track down what part of this is causing the github action to fail?

**Comment by hrshdhgd** [comment]
https://github.com/monarch-initiative/semsimian-server/compare/5e9d5e1...3638cf2

**Comment by justaddcoffee** [comment]
> [5e9d5e1...3638cf2](https://github.com/monarch-initiative/semsimian-server/compare/5e9d5e1...3638cf2)

Hmm, weird. I see these changes when I compare these commits on the command line:

```
~/PythonProject/semsimian-server add-semsimian-metric $ git diff  5e9d5e1..3638cf2 
diff --git a/.github/workflows/test.yaml b/.github/workflows/test.yaml
index 9d4d49a..6531060 100644
--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -1,9 +1,9 @@
 name: Test SemSimian Server
 on:
   push:
-    branches: [ main ]
+    branches: [ main, fix_gh_actions, test_build ]
   pull_request:
-    branches: [ main ]
+    branches: [ main, fix_gh_actions, test_build ]
 
 jobs:
   test:
diff --git a/Cargo.lock b/Cargo.lock
index dee1e0e..9d5d720 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1527,8 +1527,9 @@ checksum = "94143f37725109f92c262ed2cf5e59bce7498c01bcc1502d7b9afe439a4e9f49"
 
 [[package]]
 name = "semsimian"
-version = "0.2.14"
-source = "git+https://github.com/monarch-initiative/semsimian.git#b63a395198a2e4f5847eb9c9ffc538b3d4555035"
+version = "0.2.12"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "c6104e5c7b6b756735ad1febd80a4f87130ebdf639cf82c7264e10554fe81d55"
 dependencies = [
  "cargo-llvm-cov",
  "csv",
diff --git a/Cargo.toml b/Cargo.toml
index acc8ebc..ab9560e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,7 +13,7 @@ name = "semserver"
 path = "src/bin.rs"
 
 [dependencies]
-semsimian = { git = "https://github.com/monarch-initiative/semsimian.git", commit= "b63a395" }
+semsimian = ">=0.2.11"
 rocket = {version = "0.5.0-rc.2", features = ["json"]}
 serde = "*"
 chrono = ">=0.4.31"
diff --git a/src/bin.rs b/src/bin.rs
index ac2c94a..9b10222 100644
--- a/src/bin.rs
+++ b/src/bin.rs
@@ -7,8 +7,8 @@ use semsimian_server::{compare_termsets, search, say_hello};
 
 #[launch]
 pub fn rocket() -> _ {
-    compare_termsets("HP:0000001,HP:0000002", "HP:0000003,HP:0000004", "AncestorInformationContent");
-    search("HP:0000001,HP:0000002", "ZFIN", "AncestorInformationContent", Some(1));
+    compare_termsets("HP:0000001,HP:0000002", "HP:0000003,HP:0000004");
+    search("HP:0000001,HP:0000002", "ZFIN", Some(1));
     rocket::build().mount("/", routes![
         say_hello,
         compare_termsets,
diff --git a/src/lib.rs b/src/lib.rs
index e6bf140..997f1e7 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -6,7 +6,6 @@ use std::sync::Mutex;
 
 use lazy_static::lazy_static;
 use rocket::serde::json::Json;
-use semsimian::enums::MetricEnum;
 use semsimian::enums::SearchTypeEnum;
 use semsimian::termset_pairwise_similarity::{
     TermsetPairwiseSimilarity as Tsps, TermsetPairwiseSimilarity,
@@ -27,8 +26,8 @@ pub fn say_hello() -> &'static str {
     "Semsimian Server Online"
 }
 
-#[get("/compare/<termset1>/<termset2>/<metric>")]
-pub fn compare_termsets(termset1: &str, termset2: &str, metric: &str) -> Json<Tsps> {
+#[get("/compare/<termset1>/<termset2>")]
+pub fn compare_termsets(termset1: &str, termset2: &str) -> Json<Tsps> {
     // split termset1 and termset2 into vectors of TermIDs
     let mut terms1: HashSet<TermID> = HashSet::new();
     for term in termset1.split(",") {
@@ -45,17 +44,14 @@ pub fn compare_termsets(termset1: &str, termset2: &str, metric: &str) -> Json<Ts
         \n",
         terms1, terms2
     );
-    let metric: MetricEnum = MetricEnum::from_string(&Some(metric)).unwrap();
-    let result =
-        RSS.termset_pairwise_similarity(&terms1, &terms2, &metric);
+    let result = RSS.termset_pairwise_similarity(&terms1, &terms2);
     Json(result)
 }
 
-#[get("/search/<termset>/<prefix>/<metric>?<limit>")]
+#[get("/search/<termset>/<prefix>?<limit>")]
 pub fn search(
     termset: &str,
     prefix: &str,
-    metric: &str,
     limit: Option<usize>,
 ) -> Json<Vec<(f64, Option<TermsetPairwiseSimilarity>, TermID)>> {
     let assoc_predicate: HashSet<TermID> = HashSet::from(["biolink:has_phenotype".to_string()]);
@@ -77,9 +73,7 @@ pub fn search(
         &None,
         &subject_prefixes,
         &search_type,
-        // convert metric string to MetricEnum using from_string respecting Option expected by the from_string function
-        &MetricEnum::from_string(&Some(metric)).unwrap_or(MetricEnum::AncestorInformationContent),
-        Some(limit),
+        Some(limit)
     );
     println!("Result - {:?}", result);
 
diff --git a/tests/test_server.rs b/tests/test_server.rs
index 5d0721a..3280ade 100644
--- a/tests/test_server.rs
+++ b/tests/test_server.rs
@@ -3,7 +3,7 @@ use semsimian_server::{compare_termsets, search};
 // test compare_termsets function
 #[test]
 fn test_compare() {
-    let response = compare_termsets(&*"MP:0010771".to_string(), &*"HP:0004325".to_string(), "AncestorInformationContent");
+    let response = compare_termsets(&*"MP:0010771".to_string(), &*"HP:0004325".to_string());
     let tsps = &response.0;
     let integument_phenotype = "MP:0010771";
     let expected_first_match = String::from("match_source");
@@ -30,7 +30,6 @@ fn test_search() {
     let _response = search(
         &*"HP:0000001,HP:0000002".to_string(),
         &*"ZFIN".to_string(),
-        "AncestorInformationContent",
         Some(1),
     );
 }
\ No newline at end of file
```

**Comment by caufieldjh** [comment]
I can also build locally with commit 3638cf2c8755b21d28486ada0c5d2ffdf916a9aa
and cargo test with RUST_BACKTRACE=1 is:
```rust
    Finished test [unoptimized + debuginfo] target(s) in 0.13s
     Running unittests src/lib.rs (target/debug/deps/semsimian_server-c05c76864fae387b)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/bin.rs (target/debug/deps/semserver-b2e913ef22aa22a7)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/test_server.rs (target/debug/deps/test_server-9c9022925049fefd)

running 2 tests
test test_compare ... FAILED
test test_search ... FAILED

failures:

---- test_compare stdout ----
thread 'test_compare' panicked at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.12/src/lib.rs:99:29:
Resource returned nothing with predicates: no such table: entailed_edge
stack backtrace:
   0: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   2: semsimian::RustSemsimian::new::{{closure}}
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.12/src/lib.rs:99:29
   3: core::option::Option<T>::unwrap_or_else
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/option.rs:976:21
   4: semsimian::RustSemsimian::new
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.12/src/lib.rs:93:19
   5: semsimian_server::utils::get_rss_instance
             at ./src/utils.rs:49:31
   6: <semsimian_server::RSS as core::ops::deref::Deref>::deref::__static_ref_initialize
             at ./src/lib.rs:19:37
   7: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
   8: lazy_static::lazy::Lazy<T>::get::{{closure}}
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/inline_lazy.rs:31:29
   9: std::sync::once::Once::call_once::{{closure}}
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sync/once.rs:149:41
  10: std::sys_common::once::futex::Once::call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/once/futex.rs:124:21
  11: std::sync::once::Once::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sync/once.rs:149:9
  12: lazy_static::lazy::Lazy<T>::get
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/inline_lazy.rs:30:9
  13: <semsimian_server::RSS as core::ops::deref::Deref>::deref::__stability
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/lib.rs:142:21
  14: <semsimian_server::RSS as core::ops::deref::Deref>::deref
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/lib.rs:144:17
  15: semsimian_server::compare_termsets
             at ./src/lib.rs:47:18
  16: test_server::test_compare
             at ./tests/test_server.rs:6:20
  17: test_server::test_compare::{{closure}}
             at ./tests/test_server.rs:5:18
  18: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- test_search stdout ----
thread 'test_search' panicked at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.12/src/lib.rs:99:29:
Resource returned nothing with predicates: no such table: entailed_edge
stack backtrace:
   0: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   2: semsimian::RustSemsimian::new::{{closure}}
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.12/src/lib.rs:99:29
   3: core::option::Option<T>::unwrap_or_else
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/option.rs:976:21
   4: semsimian::RustSemsimian::new
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.12/src/lib.rs:93:19
   5: semsimian_server::utils::get_rss_instance
             at ./src/utils.rs:49:31
   6: <semsimian_server::RSS_MUTEX as core::ops::deref::Deref>::deref::__static_ref_initialize
             at ./src/lib.rs:20:61
   7: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
   8: lazy_static::lazy::Lazy<T>::get::{{closure}}
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/inline_lazy.rs:31:29
   9: std::sync::once::Once::call_once::{{closure}}
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sync/once.rs:149:41
  10: std::sys_common::once::futex::Once::call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/once/futex.rs:124:21
  11: std::sync::once::Once::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sync/once.rs:149:9
  12: lazy_static::lazy::Lazy<T>::get
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/inline_lazy.rs:30:9
  13: <semsimian_server::RSS_MUTEX as core::ops::deref::Deref>::deref::__stability
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/lib.rs:142:21
  14: <semsimian_server::RSS_MUTEX as core::ops::deref::Deref>::deref
             at /home/harry/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lazy_static-1.4.0/src/lib.rs:144:17
  15: semsimian_server::search
             at ./src/lib.rs:69:18
  16: test_server::test_search
             at ./tests/test_server.rs:30:21
  17: test_server::test_search::{{closure}}
             at ./tests/test_server.rs:29:17
  18: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    test_compare
    test_search

test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.15s

error: test failed, to rerun pass `--test test_server`
```

**Comment by glass-ships** [comment]
I've been making incremental changes locally and trying to build in between, and what i've found is: 

- Updating semsimian to "=0.2.14-rc.3" fails to build, because calls for `associations_search()` and `termset_pairwise_similarity()` require `metric` as an argument (as expected)

- If I add `&MetricEnum::AncestorInformationContent` directly as an argument, it builds without issue, something like: 
```rust
#[get("/compare/<termset1>/<termset2>/<metric>")]
pub fn compare_termsets(termset1: &str, termset2: &str, metric: Option<&str>) -> Json<Tsps> {
    ...
    let result = RSS.termset_pairwise_similarity(&terms1, &terms2, &MetricEnum::AncestorInformationContent);
    Json(result)
```


- if i then add "metric" as a parameter/argument, something like:
```rust
#[get("/compare/<termset1>/<termset2>/<metric>")]
pub fn compare_termsets(termset1: &str, termset2: &str, metric: Option<&str>) -> Json<Tsps> {
    ...
    let result = RSS.termset_pairwise_similarity(
        &terms1,
        &terms2,
        &MetricEnum::from_string(&metric).unwrap_or(MetricEnum::AncestorInformationContent),
    );
    Json(result)
}
```
This is what introduces the failure to build that we keep seeing.  
My minimal Rust knowledge is unfortunately not much use beyond this, but could it be something to do with the `from_string` business we're trying to do?

**Comment by hrshdhgd** [comment]
That's closer than any of us have gotten so great investigation @glass-ships ! `from_string` is something I wrote as a `python` => `rust` conduit such that passing a string from the `python` side equates to an enum value on the `rust` side.

Since you are completely in `rust`. You should directly use `&MetricEnum::whatever-metric-you-want` instead of `&str` as `metric`. I think this may solve the issue

**Comment by glass-ships** [comment]
the only question I'd have about that is: will Rust/Rocket know how to go from a string in the url to an enum value? 
for example, a server request might look like: 
`<SERVER_URL>/compare/HP:000001/HP:000002/ancestor_information_content`
Wouldn't it need to get converted at some point?

**Comment by hrshdhgd** [comment]
I think I know what to do ... experimenting

**Comment by hrshdhgd** [comment]
```
impl<'a> FromParam<'a> for MetricEnum {
    type Error = &'a str;

    fn from_param(param: &'a str) -> Result<Self, Self::Error> {
        match param {
            "jaccard_similarity" => Ok(MetricEnum::JaccardSimilarity),
            "phenodigm_score" => Ok(MetricEnum::PhenodigmScore),
            "cosine_similarity" => Ok(MetricEnum::CosineSimilarity),
            _ | None => Ok(MetricEnum::AncestorInformationContent),
        }
    }
}
```

This needs to be in your project. The question is where?

**Comment by hrshdhgd** [comment]
That did not resolve the issue......

**Comment by hrshdhgd** [comment]
@glass-ships , @caufieldjh : could either of you run the final commit locally to see if it works? GH Actions take a long time

**Comment by hrshdhgd** [comment]
I think this is it!!!

**Comment by glass-ships** [comment]
that it does - seems to be working for me though! i'll try testing with some various values for metric as well, but it builds, tests pass, and server runs without issue - good find on the `from_param` business!

**Comment by caufieldjh** [comment]
This also builds for me, and tests fail, but I realize this is because of the same error I usually get with freshly installed semsimian: `Resource returned nothing with predicates: no such table: entailed_edge`
So I will consider this a success.

**Comment by hrshdhgd** [comment]
I'm making a small edit for the code to see if it works. If this works, we'll have one common place to map the string to enum value conversion (in semsimian) and not in 2 places.

**Comment by glass-ships** [comment]
it's worth noting that `cosine_similarity` throws a 500 error:
```
thread 'rocket-worker-thread' panicked at 'called `Result::unwrap()` on an `Err` value: "Metric not implemented for termset comparison"', /home/glass/.cargo/registry/src/index.crates.io-6f17d22bba15001f/semsimian-0.2.14-rc.3/src/lib.rs:507:14
```
but that ... might be expected? @kevinschaper may be able to weigh in on that

**Comment by hrshdhgd** [comment]
cosine is not implemented... That is expected

**Comment by hrshdhgd** [comment]
> I'm making a small edit for the code to see if it works. If this works, we'll have one common place to map the string to enum value conversion (in semsimian) and not in 2 places.

Did not work. For now let's keep the solution that works.

**Review by glass-ships — APPROVED** [review]

**Pull request merged**